### PR TITLE
Feature tidy up

### PR DESCRIPTION
**Pull request by EdAtkin**
# Pull request description
Mainly tidying up some bits of code especially the FarDetectorStruct class. Removing allocation of lots of arrays on the heap to use std::vectors instead. Also a general bit of tidy up here and there.

## Changes or fixes
Rename of FDMC struct to FarDetectorStruct. Also moving many of the members of this struct to be vectors. Small clean up of splineFDBase class.

## Examples
e.g. int* modes = new int[nEvents]; -> std::vector<int> modes; modes.resize(nEvents);

### CODE REVIEW

File: samplePDF/FarDetectorCoreInfoStruct.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  std::vector<double> total_w;
+
+  std::vector<int> XBin;
+  std::vector<int> YBin;
+  std::vector<int> NomXBin;
+  std::vector<int> NomYBin;
+
+  bool *isNC;
+
+  // histo pdf bins
```
Comment (KSkwarczynski) [inline samplePDF/FarDetectorCoreInfoStruct.h:49]: Why is it raw pointer while everyhitng else is vector?

Comment (luketpickering) [reply samplePDF/FarDetectorCoreInfoStruct.h:49]: Read about std::vector<bool>, it doesn't behave like other std::vector objects, or just try printing the address of the first three entries of auto myvecbool = std::vector<bool>{true,true,false};

Comment (KSkwarczynski) [reply samplePDF/FarDetectorCoreInfoStruct.h:49]: I feel dumb for not knowing this...

In this case there should be a destructor?

File: samplePDF/samplePDFFDBase.cpp [hunk 2]
```diff
@@ -785,7 +782,7 @@ void samplePDFFDBase::SetupNormParameters(){
 //A way to check whether a normalisation parameter applies to an event or not
 void samplePDFFDBase::CalcXsecNormsBins(int iSample){
 
-  fdmc_base *fdobj = &MCSamples[iSample];
+  FarDetectorCoreInfo *fdobj = &MCSamples[iSample];
```
[the remark this answers is not in the document]
Comment (luketpickering) [reply samplePDF/samplePDFFDBase.cpp:785]: I'd maybe do this anyway and see if you do copy/move anywhere. If you always construct in place with `emplace_back` or `resize`, then I think its fine/best to not have copy constructors/assignment

**Pull request merged**